### PR TITLE
Improve handling of extended operations & Icom custom operations.

### PR DESCRIPTION
**Pull request by malcolmh**
The extended & custom operations for Icom rigs have until now been realized with ad-hoc code in the rig modules and in icom.c. In order to implement a more generalized & compact method of implementing these functions, the following changes have been made:

A table-driven method has been implemented so that all extended & custom operations can be specified by the use of look-up tables in both the rig modules & in icom.c. This reduces the multiple instances of essentially the same code being endlessly repeated. Pointers to these tables have been added to the Icom priv_caps struct. This new method has been implemented & tested for the rigs that I have (IC7100, ICR8600, ICR30 & ID51). The next step will be to do this for those rig modules that call the s/get_custom_parm functions & then those functions can be deleted.

Also:

include/hamlib/rig.h: 
1. Add a new rig_conf_e member to specify a generic binary buffer type. This is similar to the STRING member, except that the buffer length needs to be included.
2. Likewise, add a binary buffer member to the value_t union.
3. Add a list of tokens to rig_caps to specify which extended operations a rig is capable of handling.

 tests/rigctl_parse.c &  tests/sprintflst.c:
Add code to handle binary buffers.

### CODE REVIEW

File: tests/rigctl_parse.c [hunk 1]
```diff
@@ -3009,6 +3030,10 @@ declare_proto_rig(get_parm)
             fprintf(fout, "%s\n", val.s);
             break;
 
+        case RIG_CONF_BINARY:
+            dump_hex((unsigned char *)buffer, val.b.l);
```
Comment (N0NB) [inline tests/rigctl_parse.c]: Hi Malcolm.  This addition of dump_hex() to rigctl_parse.c is causing a linker error on my MinGW build:

```
make[1]: Entering directory '/home/nate/builds/hamlib-4.0~git/tests'
  CC       rigctl-rigctl.o
  CC       rigctl-rigctl_parse.o
  CC       rigctl-dumpcaps.o
  CC       rigctl-sprintflst.o
  CCLD     rigctl.exe
rigctl-rigctl_parse.o: In function `rigctl_get_parm':
/home/nate/builds/hamlib-4.0~git/tests/rigctl_parse.c:3049: undefined reference to `dump_hex'
collect2: error: ld returned 1 exit status
Makefile:949: recipe for target 'rigctl.exe' failed
make[1]: *** [rigctl.exe] Error 1
make[1]: Leaving directory '/home/nate/builds/hamlib-4.0~git/tests'
Makefile:605: recipe for target 'install-recursive' failed
```

It's not easy to find out if the `dump_hex` symbol is exported into the `libhamlib.dll.a` but my check for the string shows it to not be found while the individual `debug.o` file does show the string.  I know just enough about MinGW sometimes to get things working but often I am stumped.  Obviously `dump_hex` is available on Linux as rigctl links without issue, so this is just a MinGW issue at the moment.  I will note that `dump_hex` is used in the barret backend and in iofunc.c.

Mike, do you have an idea here?  If we can't resolve this, perhaps the `dump_hex` will need to be done another way.

**Comment by N0NB** [comment]
More digging into our declarations shows that `dump_hex` is considered an internal Hamlib function and is not an exported symbol on Windows via MinGW.  See `include/hamlib/rig_dll.h` for the export macros.

Do we make `dump_hex` a part of the exported symbols or should there be a wrapper function that is exported?

**Comment by mdblack98** [comment]
I submitted a patch for this...

 

    On Monday, January 20, 2020, 07:57:47 AM CST, Nate Bargmann <notifications@github.com> wrote:  
 
 
@N0NB requested changes on this pull request.

In tests/rigctl_parse.c:
> @@ -3009,6 +3030,10 @@ declare_proto_rig(get_parm)
             fprintf(fout, "%s\n", val.s);
             break;
 
+        case RIG_CONF_BINARY:
+            dump_hex((unsigned char *)buffer, val.b.l);

Hi Malcolm. This addition of dump_hex() to rigctl_parse.c is causing a linker error on my MinGW build:
make[1]: Entering directory '/home/nate/builds/hamlib-4.0~git/tests'
  CC       rigctl-rigctl.o
  CC       rigctl-rigctl_parse.o
  CC       rigctl-dumpcaps.o
  CC       rigctl-sprintflst.o
  CCLD     rigctl.exe
rigctl-rigctl_parse.o: In function `rigctl_get_parm':
/home/nate/builds/hamlib-4.0~git/tests/rigctl_parse.c:3049: undefined reference to `dump_hex'
collect2: error: ld returned 1 exit status
Makefile:949: recipe for target 'rigctl.exe' failed
make[1]: *** [rigctl.exe] Error 1
make[1]: Leaving directory '/home/nate/builds/hamlib-4.0~git/tests'
Makefile:605: recipe for target 'install-recursive' failed

It's not easy to find out if the dump_hex symbol is exported into the libhamlib.dll.a but my check for the string shows it to not be found while the individual debug.o file does show the string. I know just enough about MinGW sometimes to get things working but often I am stumped. Obviously dump_hex is available on Linux as rigctl links without issue, so this is just a MinGW issue at the moment. I will note that dump_hex is used in the barret backend and in iofunc.c.

Mike, do you have an idea here? If we can't resolve this, perhaps the dump_hex will need to be done another way.

**Comment by N0NB** [comment]
Mike's patch fixed the export issue and I have a clean build for both MinGW 32 and 64 bit.

**Pull request closed without merging**